handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -0,0 +1,1951 @@
+            int bytesConsumed, int bytesProduced, SSLEngineResult.HandshakeStatus status) throws SSLException {
+        return new SSLEngineResult(
+                getEngineStatus(), mayFinishHandshake(status != FINISHED ? getHandshakeStatus() : status)
+                , bytesConsumed, bytesProduced);
+    }
+
+    private void closeAll() throws SSLException {
+        receivedShutdown = true;
+        closeOutbound();
+        closeInbound();
+    }
+
+    private void rejectRemoteInitiatedRenegation() throws SSLHandshakeException {
+        if (rejectRemoteInitiatedRenegation && SSL.getHandshakeCount(ssl) > 1) {
+            // TODO: In future versions me may also want to send a fatal_alert to the client and so notify it
+            // that the renegotiation failed.
+            shutdown();
+            throw new SSLHandshakeException("remote-initiated renegotation not allowed");
+        }
+    }
+
+    public final SSLEngineResult unwrap(final ByteBuffer[] srcs, final ByteBuffer[] dsts) throws SSLException {
+        return unwrap(srcs, 0, srcs.length, dsts, 0, dsts.length);
+    }
+
+    private ByteBuffer[] singleSrcBuffer(ByteBuffer src) {
+        singleSrcBuffer[0] = src;
+        return singleSrcBuffer;
+    }
+
+    private void resetSingleSrcBuffer() {
+        singleSrcBuffer[0] = null;
+    }
+
+    private ByteBuffer[] singleDstBuffer(ByteBuffer src) {
+        singleDstBuffer[0] = src;
+        return singleDstBuffer;
+    }
+
+    private void resetSingleDstBuffer() {
+        singleDstBuffer[0] = null;
+    }
+
+    @Override
+    public final synchronized SSLEngineResult unwrap(
+            final ByteBuffer src, final ByteBuffer[] dsts, final int offset, final int length) throws SSLException {
+        try {
+            return unwrap(singleSrcBuffer(src), 0, 1, dsts, offset, length);
+        } finally {
+            resetSingleSrcBuffer();
+        }
+    }
+
+    @Override
+    public final synchronized SSLEngineResult wrap(ByteBuffer src, ByteBuffer dst) throws SSLException {
+        try {
+            return wrap(singleSrcBuffer(src), dst);
+        } finally {
+            resetSingleSrcBuffer();
+        }
+    }
+
+    @Override
+    public final synchronized SSLEngineResult unwrap(ByteBuffer src, ByteBuffer dst) throws SSLException {
+        try {
+            return unwrap(singleSrcBuffer(src), singleDstBuffer(dst));
+        } finally {
+            resetSingleSrcBuffer();
+            resetSingleDstBuffer();
+        }
+    }
+
+    @Override
+    public final synchronized SSLEngineResult unwrap(ByteBuffer src, ByteBuffer[] dsts) throws SSLException {
+        try {
+            return unwrap(singleSrcBuffer(src), dsts);
+        } finally {
+            resetSingleSrcBuffer();
+        }
+    }
+
+    @Override
+    public final Runnable getDelegatedTask() {
+        // Currently, we do not delegate SSL computation tasks
+        // TODO: in the future, possibly create tasks to do encrypt / decrypt async
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)